File: capstone/Backend/assesment-portal/src/main/java/com/project/assesmentportal/services/CategoryService.java
package com.project.assesmentportal.services;

import java.util.List;

import com.project.assesmentportal.dto.ApiResponse;
import com.project.assesmentportal.dto.CategoryDto;
import com.project.assesmentportal.dto.QuizDto;

/**
 * Service interface for managing category-related operations.
 */
public interface CategoryService {

    /**
     * Adds a new category.
     * @param categoryDto The CategoryDto representing the category to be added.
     * @return The String message for the added category.
     */
    ApiResponse addCategory(CategoryDto categoryDto);

    /**
     * Retrieves a list of all categories.
     * @return A list of CategoryDto objects representing all categories.
     */
    List<CategoryDto> getCategories();

    /**
     * Retrieves a category by its ID.
     * @param categoryId The ID of the category to retrieve.
     * @return The CategoryDto of the retrieved category.
     */
    CategoryDto getCategoryById(long categoryId);

    /**
     * Updates a category.
     * @param categoryDto The updated CategoryDto.
     * @param categoryId  The ID of the category to update.
     * @return String message for the updated category.
     */
    ApiResponse updateCategory(CategoryDto categoryDto, long categoryId);

    /**
     * Deletes a category by its ID.
     * @param categoryId The ID of the category to delete.
     */
    ApiResponse deleteCategory(long categoryId);

    /**
     * Gets list of quizzes for a category.
     * @param categoryId if of the category.
     * @return list of quizzes.
     */
    List<QuizDto> getQuizzesByCategory(long categoryId);
}

File: capstone/Frontend/src/Services/CategoryService.js
import axios from 'axios';

const CATEGORY_BASE_URL = 'http://localhost:8080/categories';

class CategoryService{

    addCategory(Category){
        return axios.post(CATEGORY_BASE_URL, Category);
    }

    getAllCategories(){
        return axios.get(CATEGORY_BASE_URL);
    }

    getCategoryById(id){
        return axios.get(CATEGORY_BASE_URL+'/'+id);
    }

    updateCategory(id, Category){
        return axios.put(CATEGORY_BASE_URL+'/'+id,Category);
    }

    deleteCategory(id){
        return axios.delete(CATEGORY_BASE_URL+'/'+id);
    } 

    getQuizzesByCategory(id){
        return axios.get(CATEGORY_BASE_URL+'/'+id+'/'+"quizzes");
    }
}

export default new CategoryService();